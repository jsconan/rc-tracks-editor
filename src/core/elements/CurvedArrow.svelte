<script>
    // Licensed under GNU Public License version 3
    // Copyright (c) 2022 Jean-SÃ©bastien CONAN

    import { SVGPath, Vector2D } from '../models';
    import { attributeList, enlargeArc } from '../helpers';

    export let cx = 0;
    export let cy = 0;
    export let width = 1;
    export let thickness = void 0;
    export let clockwise = true;
    export let radius = 1;
    export let angle = 1;
    export let start = 0;
    export let fill = void 0;
    export let stroke = void 0;
    export let transform = void 0;

    /**
     * Builds the SVG path for rendering a curved arrow.
     * @param {number} arrowWidth - The width of the arrow.
     * @param {number} arrowThickness - The thickness of the arrow.
     * @param {boolean} arrowClockwise - Is the arrow clockwise or counter-clockwise?
     * @param {number} curveRadius - The radius of the curve.
     * @param {number} curveAngle - Tha angle of the curve.
     * @param {number} startAngle - The start angle of the curve.
     * @param {number} centerX - The X-coordinate of the center of the curve.
     * @param {number} centerY - The Y-coordinate of the center of the curve.
     * @private
     */
    function curvedArrowPath(
        arrowWidth,
        arrowThickness,
        arrowClockwise,
        curveRadius,
        curveAngle,
        startAngle,
        centerX,
        centerY
    ) {
        const center = new Vector2D(centerX, centerY);

        const tipEdgeRadius = curveRadius;
        const tipInnerRadius = tipEdgeRadius - arrowWidth / 2;
        const tipOuterRadius = tipInnerRadius + arrowWidth;
        const innerRadius = tipEdgeRadius - arrowThickness / 2;
        const outerRadius = innerRadius + arrowThickness;
        const bodyAngle = enlargeArc(curveAngle, tipOuterRadius, -arrowWidth);

        let bodyStart, bodyEnd, tipStart, tipEnd;

        if (arrowClockwise) {
            bodyStart = startAngle;
            bodyEnd = bodyStart + bodyAngle;
            tipStart = bodyEnd;
            tipEnd = bodyStart + curveAngle;
        } else {
            bodyEnd = startAngle + curveAngle;
            bodyStart = bodyEnd - bodyAngle;
            tipEnd = startAngle;
            tipStart = bodyStart;
        }

        const path = new SVGPath();

        if (arrowClockwise) {
            return path
                .polarMoveTo(innerRadius, bodyStart, center)
                .arcCurveTo(innerRadius, -bodyAngle, Vector2D.polar(innerRadius, bodyEnd, center))
                .polarLineTo(tipInnerRadius, tipStart, center)
                .polarLineTo(tipEdgeRadius, tipEnd, center)
                .polarLineTo(tipOuterRadius, tipStart, center)
                .polarLineTo(outerRadius, bodyEnd, center)
                .arcCurveTo(outerRadius, bodyAngle, Vector2D.polar(outerRadius, bodyStart, center))
                .close();
        } else {
            return path
                .polarMoveTo(tipEdgeRadius, tipEnd, center)
                .polarLineTo(tipOuterRadius, tipStart, center)
                .polarLineTo(outerRadius, bodyStart, center)
                .arcCurveTo(outerRadius, -bodyAngle, Vector2D.polar(outerRadius, bodyEnd, center))
                .polarLineTo(innerRadius, bodyEnd, center)
                .arcCurveTo(innerRadius, bodyAngle, Vector2D.polar(innerRadius, bodyStart, center))
                .polarLineTo(tipInnerRadius, tipStart, center)
                .close();
        }
    }

    $: if ('undefined' === typeof thickness) {
        thickness = width / 3;
    }
</script>

<path
    d={curvedArrowPath(width, thickness, clockwise, radius, angle, start, cx, cy)}
    {...attributeList(fill, 'fill')}
    {...attributeList(stroke, 'stroke')}
    {transform}
/>
