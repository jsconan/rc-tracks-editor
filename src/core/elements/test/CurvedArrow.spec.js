/**
 * RC Tracks Editor
 * Copyright (c) 2022 Jean-SÃ©bastien CONAN
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import { render } from '@testing-library/svelte';
import { tick } from 'svelte';
import CurvedArrow from '../CurvedArrow.svelte';

describe('CurvedArrow', () => {
    it('renders with default values', () => {
        const { container } = render(CurvedArrow);

        expect(container).toMatchSnapshot();
    });

    it('renders with the given parameters', () => {
        const props = {
            cx: 100,
            cy: 150,
            rotation: 60,
            height: 50,
            radius: 50,
            angle: 30,
            thickness: 25,
            clockwise: true,
            fill: 'red',
            stroke: {
                stroke: 'green',
                'stroke-width': 2
            },
            transform: 'rotate(90)'
        };
        const { container } = render(CurvedArrow, { props });

        expect(container).toMatchSnapshot();
    });

    it.each([
        ['cx', { cx: 40 }],
        ['cy', { cy: 40 }],
        ['rotation', { rotation: 90 }],
        ['width', { width: 60 }],
        ['height', { height: 60 }],
        ['radius', { radius: 60 }],
        ['angle', { angle: 90 }],
        ['thickness', { thickness: 50 }],
        ['clockwise', { clockwise: false }],
        ['fill', { fill: 'blue' }],
        ['stroke', { stroke: 'blue' }],
        ['stroke-width', { stroke: { 'stroke-width': 3 } }],
        ['transform', { transform: 'rotate(180)' }]
    ])('updates when the parameter %s is modified', async (title, update) => {
        const props = {
            cx: 100,
            cy: 150,
            rotation: 60,
            width: 50,
            height: 50,
            radius: 50,
            thickness: 25,
            clockwise: true,
            fill: 'red',
            stroke: {
                stroke: 'green',
                'stroke-width': 2
            },
            transform: 'rotate(90)'
        };
        const rendered = render(CurvedArrow, { props });

        await tick();
        rendered.component.$set(update);
        await tick();
        expect(rendered.container).toMatchSnapshot();
    });
});
