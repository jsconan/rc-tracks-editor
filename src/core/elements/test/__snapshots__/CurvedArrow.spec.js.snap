// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurvedArrow renders with default values 1`] = `
<body>
  <div>
    <path
      d="M 0.8333333333333334,0 A 0.8333333333333334,0.8333333333333334 0 0 0 0.6637996728935979,-0.5037999987205211 L 0.3982798037361587,-0.30227999923231264 L 0.9998476951563913,0.01745240643728351 L 1.1948394112084761,-0.9068399976969379 L 0.929319542051037,-0.7053199982087295 A 1.1666666666666667,1.1666666666666667 0 0 1 1.1666666666666667,0 Z"
    />
  </div>
</body>
`;

exports[`CurvedArrow renders with the given parameters 1`] = `
<body>
  <div>
    <path
      d="M 118.75,182.47595264191645 A 37.5,37.5 0 0 0 123.18886761511514,179.47077227913556 L 115.45924507674343,169.6471815194237 L 100,200 L 146.3777352302303,208.9415445582711 L 138.64811269185856,199.11795379855926 A 62.5,62.5 0 0 1 131.25,204.12658773652743 Z"
      fill="red"
      stroke="green"
      stroke-width="2"
      transform="rotate(90)"
    />
  </div>
</body>
`;

exports[`CurvedArrow updates when the parameter angle is modified 1`] = `
<body>
  <div>
    <path
      d="M 118.75,182.47595264191645 A 37.5,37.5 0 0 1 86.07199634467996,184.81753457925174 L 90.71466422978665,173.21168971950118 L 56.69872981077806,175 L 72.14399268935993,219.6350691585035 L 76.78666057446661,208.02922429875292 A 62.5,62.5 0 0 0 131.25,204.12658773652743 Z"
      fill="red"
      stroke="green"
      stroke-width="2"
      transform="rotate(90)"
    />
  </div>
</body>
`;

exports[`CurvedArrow updates when the parameter clockwise is modified 1`] = `
<body>
  <div>
    <path
      d="M 125,193.30127018922192 L 89.30647539976758,224.23374254087037 L 91.08872949980632,211.86145211739196 A 62.5,62.5 0 0 0 100,212.5 L 100,187.5 A 37.5,37.5 0 0 1 94.65323769988379,187.11687127043518 L 96.43549179992253,174.74458084695678 Z"
      fill="red"
      stroke="green"
      stroke-width="2"
      transform="rotate(90)"
    />
  </div>
</body>
`;

exports[`CurvedArrow updates when the parameter cx is modified 1`] = `
<body>
  <div>
    <path
      d="M 58.75,182.47595264191645 A 37.5,37.5 0 0 0 63.18886761511514,179.47077227913556 L 55.45924507674343,169.6471815194237 L 40,200 L 86.37773523023029,208.9415445582711 L 78.64811269185857,199.11795379855926 A 62.5,62.5 0 0 1 71.25,204.12658773652743 Z"
      fill="red"
      stroke="green"
      stroke-width="2"
      transform="rotate(90)"
    />
  </div>
</body>
`;

exports[`CurvedArrow updates when the parameter cy is modified 1`] = `
<body>
  <div>
    <path
      d="M 118.75,72.47595264191645 A 37.5,37.5 0 0 0 123.18886761511514,69.47077227913555 L 115.45924507674343,59.6471815194237 L 100,90 L 146.3777352302303,98.94154455827109 L 138.64811269185856,89.11795379855926 A 62.5,62.5 0 0 1 131.25,94.12658773652741 Z"
      fill="red"
      stroke="green"
      stroke-width="2"
      transform="rotate(90)"
    />
  </div>
</body>
`;

exports[`CurvedArrow updates when the parameter fill is modified 1`] = `
<body>
  <div>
    <path
      d="M 118.75,182.47595264191645 A 37.5,37.5 0 0 0 123.18886761511514,179.47077227913556 L 115.45924507674343,169.6471815194237 L 100,200 L 146.3777352302303,208.9415445582711 L 138.64811269185856,199.11795379855926 A 62.5,62.5 0 0 1 131.25,204.12658773652743 Z"
      fill="blue"
      stroke="green"
      stroke-width="2"
      transform="rotate(90)"
    />
  </div>
</body>
`;

exports[`CurvedArrow updates when the parameter radius is modified 1`] = `
<body>
  <div>
    <path
      d="M 123.75,191.13620667976085 A 47.5,47.5 0 0 0 126.35745509452636,189.51625691965305 L 119.42128270122994,179.117241940797 L 100,210 L 147.16597227441557,220.71330185622122 L 140.22979988111916,210.31428687736516 A 72.5,72.5 0 0 1 136.25,212.7868417743718 Z"
      fill="red"
      stroke="green"
      stroke-width="2"
      transform="rotate(90)"
    />
  </div>
</body>
`;

exports[`CurvedArrow updates when the parameter start is modified 1`] = `
<body>
  <div>
    <path
      d="M 100,187.5 A 37.5,37.5 0 0 0 105.34676230011621,187.11687127043518 L 103.56450820007748,174.74458084695678 L 75.00000000000001,193.30127018922195 L 110.69352460023244,224.23374254087037 L 108.9112705001937,211.86145211739196 A 62.5,62.5 0 0 1 100,212.5 Z"
      fill="red"
      stroke="green"
      stroke-width="2"
      transform="rotate(90)"
    />
  </div>
</body>
`;

exports[`CurvedArrow updates when the parameter stroke is modified 1`] = `
<body>
  <div>
    <path
      d="M 118.75,182.47595264191645 A 37.5,37.5 0 0 0 123.18886761511514,179.47077227913556 L 115.45924507674343,169.6471815194237 L 100,200 L 146.3777352302303,208.9415445582711 L 138.64811269185856,199.11795379855926 A 62.5,62.5 0 0 1 131.25,204.12658773652743 Z"
      fill="red"
      stroke="blue"
      transform="rotate(90)"
    />
  </div>
</body>
`;

exports[`CurvedArrow updates when the parameter stroke-width is modified 1`] = `
<body>
  <div>
    <path
      d="M 118.75,182.47595264191645 A 37.5,37.5 0 0 0 123.18886761511514,179.47077227913556 L 115.45924507674343,169.6471815194237 L 100,200 L 146.3777352302303,208.9415445582711 L 138.64811269185856,199.11795379855926 A 62.5,62.5 0 0 1 131.25,204.12658773652743 Z"
      fill="red"
      stroke-width="3"
      transform="rotate(90)"
    />
  </div>
</body>
`;

exports[`CurvedArrow updates when the parameter thickness is modified 1`] = `
<body>
  <div>
    <path
      d="M 112.5,171.65063509461098 A 25,25 0 0 0 115.45924507674343,169.6471815194237 L 100,200 L 146.3777352302303,208.9415445582711 A 75,75 0 0 1 137.5,214.9519052838329 Z"
      fill="red"
      stroke="green"
      stroke-width="2"
      transform="rotate(90)"
    />
  </div>
</body>
`;

exports[`CurvedArrow updates when the parameter transform is modified 1`] = `
<body>
  <div>
    <path
      d="M 118.75,182.47595264191645 A 37.5,37.5 0 0 0 123.18886761511514,179.47077227913556 L 115.45924507674343,169.6471815194237 L 100,200 L 146.3777352302303,208.9415445582711 L 138.64811269185856,199.11795379855926 A 62.5,62.5 0 0 1 131.25,204.12658773652743 Z"
      fill="red"
      stroke="green"
      stroke-width="2"
      transform="rotate(180)"
    />
  </div>
</body>
`;

exports[`CurvedArrow updates when the parameter width is modified 1`] = `
<body>
  <div>
    <path
      d="M 118.75,182.47595264191645 A 37.5,37.5 0 0 0 123.18886761511514,179.47077227913556 L 115.45924507674343,169.6471815194237 L 100,200 L 146.3777352302303,208.9415445582711 L 138.64811269185856,199.11795379855926 A 62.5,62.5 0 0 1 131.25,204.12658773652743 Z"
      fill="red"
      stroke="green"
      stroke-width="2"
      transform="rotate(90)"
    />
  </div>
</body>
`;
